extension-manifest-v3/src/background/utils/request.js

```diff
@@ -53,27 +53,31 @@ export default class ExtendedRequest extends Request {
     this.sourceHostname = data.sourceHostname;
   }
 
-  isFromOriginUrl(url) {
+  isFromDomain(domain) {
     const { frameAncestors } = this._originalRequestDetails;
 
-    /* Firefox APIs */
+    let url = '';
 
+    /* Firefox APIs */
     if (frameAncestors && frameAncestors.length > 0) {
-      return url === frameAncestors[frameAncestors.length - 1].url;
-    }
-
-    if (this.sourceUrl) {
-      return url === this.sourceUrl;
+      url = frameAncestors[frameAncestors.length - 1].url;
+    } else if (this.sourceUrl) {
+      url = this.sourceUrl;
+    } else {
+      /* Chrome APIs */
+
+      const { frameType, initiator } = this._originalRequestDetails;
+
+      // For frameType 'sub_frame', we can't determine the origin URL
+      // as it might be the iframe itself or any of its ancestors
+      if (frameType === 'outermost_frame' && initiator) {
+        url = initiator;
+      }
     }
 
-    /* Chrome APIs */
-
-    const { frameType, initiator } = this._originalRequestDetails;
-
-    // For frameType 'sub_frame', we can't determine the origin URL
-    // as it might be the iframe itself or any of its ancestors
-    if (frameType === 'outermost_frame' && initiator) {
-      return url.startsWith(initiator);
+    if (url) {
+      const parsedUrl = parse(url);
```

Comment:
we will likely parse the same URL over and over again. One way to do it would be to add some in-memory cache. This is what we did in V8 - https://github.com/ghostery/common/blob/main/modules/core/sources/LRU.ts